### DISCUSSION TITLE
Papis LLM

### DISCUSSION THREAD

**Discussion by isaksamsten** [papis/papis#877] [Show and tell]
Hi!

I've created a small plugin for Papis to query my library using LLMs. I call it [papis-qa](https://github.com/isaksamsten/papisqa).

It's really simple and indexes all PDFs in a library using OpenAI embeddings and construct answers using GPT4o. It simply uses [paper-qa](https://github.com/whitead/paper-qa) for embedding and question/answering.

A small example from my own library:

```
$ papis ai index
```

This will index all papers in my default library and store the embeddings in a file (inside the configuration directory). Total cost 0.6$. Every time a new document is added we can rerun index and it will only index the new (or changed) documents.

Next we can ask questions about the library:

```
$ papis ai ask "What is the difference between multirocket and Rocket?"
```

In my library one answer could be:

```
# Question
What is the difference between multirocket and Rocket?

# Answer
MultiRocket expands upon Rocket by incorporating several key enhancements.
Unlike Rocket, which applies global max pooling and 'proportion of positive
values' (PPV) pooling, MultiRocket introduces three additional pooling
operations (Dempster et al., 2023 pages 6-7; Tan et al., 2022 pages 2-3).
MultiRocket transforms both the original time series and their first-order
difference, capturing rate of change (Tan et al., 2022 pages 7-7). It utilizes
the same 84 fixed kernels as MiniRocket but applies them to both the original
and differenced time series, resulting in broader and more nuanced feature
extraction (Tan et al., 2022 pages 7-8). These enhancements allow MultiRocket
to achieve higher accuracy with significantly improved speed, positioning it as
a state-of-the-art time series classification algorithm (Dempster et al., 2023
pages 6-7).

# References
 - Dempster et al., 2023 pages 6-7
 - Tan et al., 2022 pages 2-3
 - Tan et al., 2022 pages 7-8
 - Tan et al., 2022 pages 7-7
 - Tan et al., 2022 pages 6-7
```

You could try it your self by injecting the plugin into your `pipx` environment or install it using `pip` if you are not using `pipx`.

So far this is more of an experiment, but perhaps it can be useful for someone more than me :-)

Best,
Isak

**Comment by jghauser** [comment]
This looks very interesting! Thanks a lot for this very useful addition to the Papis ecosystem :heart: 

EDIT: can we add it to the README here? It will have more visibility that in a discussion post :)

**Reply by isaksamsten** [reply]
Sure! Of course you can add it to the README, but please note that it is rather inflexible. If the library changes old files will still be indexed and so on. I'll work on improving it but it is actually rather useful already

**Reply by jghauser** [reply]
Ok, good to know! It really does look useful. It might be the thing that gets me to finally sign up for chatgpt haha

**Reply by isaksamsten** [reply]
paper-qa supports locally hosted models so it should be possible to implement that for those who don't want to pay for or be subject to OpenAI privacy issues. For me the privacy issue is irrelevant for my paper collection :-)

**Reply by alexfikl** [reply]
> EDIT: can we add it to the README here? It will have more visibility that in a discussion post :)

Agreed that we should add it to the README! @isaksamsten Feel free to make a pull request to add it to the little table at the end there when you feel like it's in a good enough shape! :1st_place_medal:
